### PR TITLE
[dapp-kit] default storage to memory storage when local storage isn't available in an SSR context

### PR DESCRIPTION
**Pull request by williamrobertson13**
## Description 

dapp-kit breaks when used in a framework like Next since we default to using local storage for persisting connection data which doesn't exist in an SSR context. While we're working on better support for these frameworks, this PR fixes the immediate breakage issue by defaulting to memory storage when local storage isn't available. I also added the option to entirely disable data persistence by setting `storage: null` in the `WalletProvider`.

Most of this was copy-pasta'd from @Jordan-Mysten's Next PR, I did hoist the storage determination logic up to the `WalletProvider` since I think it might be more clear to consumers what storage ends up being used. I don't have a super strong opinion on where this lives (`WalletProvider.tsx` or `walletStore.ts`), so I'm glad to change this back if anyone has a preference.

## Test Plan 
- Existing persistence tests pass
- Tested manually
- CI

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: sdk/dapp-kit/src/constants/walletDefaults.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import type { WalletWithRequiredFeatures } from '@mysten/wallet-standard';
+
+import { createInMemoryStore } from '../utils/stateStorage.js';
+
+export const SUI_WALLET_NAME = 'Sui Wallet';
+
+export const DEFAULT_STORAGE = typeof window === 'undefined' ? createInMemoryStore() : localStorage;
```
Comment (hayes-mysten) [inline sdk/dapp-kit/src/constants/walletDefaults.ts:10]: should we do something type `typeof window?.localStorage` here?  I know some envs will have a fake partial window object

Comment (hayes-mysten) [reply sdk/dapp-kit/src/constants/walletDefaults.ts:10]: or `window?.localStorage ?? createInMemoryStore()`

**Review by hayes-mysten — APPROVED** [review]

**Comment by XoL1507** [comment]
LGTM!

File: sdk/dapp-kit/src/constants/walletDefaults.ts (hunk 1 above)
Comment (Jordan-Mysten) [reply sdk/dapp-kit/src/constants/walletDefaults.ts:10]: If window is undefined does that identifier access throw?

Comment (hayes-mysten) [reply sdk/dapp-kit/src/constants/walletDefaults.ts:10]: yes... it looks like it does.  I expected that the `?.` would make this safe, but it does not.  probably need to do both the typeof and then the check for localStorage

Comment (williamrobertson13) [reply sdk/dapp-kit/src/constants/walletDefaults.ts:10]: I could use this bad boy, but that might be a little overkill 😆.

image

 I'll update this with the following, thanks for the call out!

```
 typeof window !== 'undefined' && window.localStorage
```

**Review by hayes-mysten — APPROVED** [review]

**Pull request merged**